### DISCUSSION TITLE
Enable users to embed or reference declarations dynamically

### DISCUSSION THREAD

**Discussion by jondavid-black** [DevOps-MBSE/AaC#664] [Ideas]
This is just a random thought, but it may actually be useful so I wanted to get it written down for discussion.

What if we effectively got rid of the concept of root?  Any schema could be a root type within a model file.  If a user decides they want to define a behavior in a stand-alone file, they can.  When they create a model that uses that behavior, they simply reference the behavior in the behavior list.  Alternatively, if there's something they just want to define in-line, they can do that too.  

So why would this be useful?  It simply enables greater reuse and may simplify model representation.  It's not difficult to imagine domain specific modeling constructs where we would want to model various execution constructs or aggregations.  In some instances it may be useful to model these in parallel.  If we can maintain a single definition of reused components, it will be simpler to manage the overall system.  For example, let's imagine a test system and an operations system.  Perhaps the test system runs within a simulation with specific characteristics that should be captured in the model.  We may not want to embed the simulation details in an operational representative model that may be used to publish deployment code or documentation, but we certainly want the operational details to be consistent within the model used for test.  There are probably a few ways this could be accomplished, but a simple way may just be to simplify reuse of things like behavior across different environment specific models.

To enable this additional flexibility, we would have to modify the basic validation to allow either an explicit declaration or reference for each non-primitive field within a schema.  For lists, each entry would be evaluated as either an explicit declaration or a reference.  We would also need to look at getting rid of the root schema and allowing any schema to be definable as a "root" in the aac file.

Perhaps this would create problems or challenges that I don't comprehend yet.  I'm curious to think through what those may be.  I'm also curious to think through the user experience.  If we did something like this, do things get easier or harder for the user?  Is the aac representation easier or harder to maintain?  Is the flexibility valuable or overly complex?  Are domain specific schemas and extensions easier or harder to utilize?  Do other ideas like implementing inheritance become more or less challenging to build or use?

**Comment by jondavid-black** [comment]
The more I think about this the more I like it.  The flexibility this would provide to enable reuse across perspectives or views could be incredibly powerful.  One example I've been considering is how the same model declarations could be used across models for representing the logical system (i.e. spec-tree aligned) and reused to represent the deployment of the system (i.e. code generation and deployment).  I may try to develop some specific examples and post here.

**Comment by jondavid-black** [comment]
A potential challenge presented here is defining the "entry point" of the model.  The models effectively establish a tree structure which has a root and various children.  Software programs have a similar challenge and use simple conventions such as defining a "main" function as the execution root.  We may need an explicit way to clearly establish that model root to enable clearer communication among people via models.  Perhaps some SysML construct such as views would be valuable to consider, or perhaps there's a simpler approach.

**Comment by cvchaparro** [comment]
I like this idea and think it would be worth discussing how it would work in a practical sense. @lizzcondrey @Coffee2Bits

**Comment by lizzcondrey** [comment]
Believe this became resolved with the merge in of #679, correct?

**Reply by Coffee2Bits** [reply]
If I understand correctly, it did.

**Reply by lizzcondrey** [reply]
@jondavid-black concur as the original poster?
